packages/alfa-rules/src/sia-r68/rule.ts

```diff
@@ -97,3 +103,37 @@ function isRequiredChild(
     };
   }
 }
+
+/**
+ * Collect all descendants of node that:
+ * * match includeNode; and
+ * * do not have an ancestor matching excludeSubtree; and
+ * * have an ancestor matching collectSubtree
+ */
+function visit<T extends Node = Node>(
+  node: Node,
+  options: Node.Traversal,
+  includeNode: Predicate<Node, T> = () => true,
+  excludeSubtree: Predicate<Node> = () => false,
+  collectSubtree: Predicate<Node> = () => true
+): Iterable<T> {
+  function* doVisit(node: Node, collect: boolean): Iterable<T> {
+    if (test<Node>(excludeSubtree, node)) {
+      return Sequence.empty();
+    }
+
+    if (collect && includeNode(node)) {
+      yield node;
+    }
+
+    collect = collect || collectSubtree(node);
+
+    const children = node.children(options);
+
+    for (const child of children) {
+      yield* doVisit(child, collect);
+    }
+  }
+
+  return doVisit(node, false);
+}
```

Comment:
This is definitely a case of premature abstraction so let's not do that 🙈 This might provide some good insights: https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction